src/libraries/System.Console/src/System/ConsolePal.iOS.cs

```diff
@@ -1,23 +1,89 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Diagnostics;
 using System.IO;
+using System.Runtime.InteropServices;
 using System.Text;
 
 namespace System
 {
     internal sealed class NSLogStream : ConsoleStream
     {
-        public NSLogStream() : base(FileAccess.Write) {}
+        private readonly StringBuilder _buffer = new StringBuilder();
+        private Decoder _decoder;
+
+        public NSLogStream() : base(FileAccess.Write)
+        {
+            _decoder = Encoding.Unicode.GetDecoder();
+        }
 
         public override int Read(Span<byte> buffer) => throw Error.GetReadNotSupported();
 
         public override unsafe void Write(ReadOnlySpan<byte> buffer)
         {
-            fixed (byte* ptr = buffer)
+            int charCount = _decoder.GetCharCount(buffer, false);
+            Span<char> charSpan = new char[charCount];
+            int count = _decoder.GetChars(buffer, charSpan, false);
+            if (count > 0)
+            {
+                WriteOrCache(_buffer, charSpan);
+            }
+        }
+
+        private static unsafe void WriteOrCache(StringBuilder cache, Span<char> charBuffer)
+        {
+            int lineStartIndex = 0;
+            for (int i = 0; i < charBuffer.Length; i++)
+            {
+                if (charBuffer[i] == '\n')
+                {
+                    Span<char> lineSpan = charBuffer.Slice(lineStartIndex, i - lineStartIndex);
+                    if (cache.Length > 0)
+                    {
+                        printer(cache.Append(lineSpan).ToString());
```

Comment:
I wonder if you can use the `StringBuilder.GetChunks()` method instead of allocating a new string here. You should be able to enumerate through the "chunks" of the StringBuilder, and write them to `Interop.Sys.Log` without allocating.

Reply:
String builder can contain more than 1 chunk so enumerating though and writing them to Interop.Sys.Log means invoking NSLog several times, which would result in the same situation as right now because NSLog automatically appends a new line to any input.